BEIMA.Backend/DeviceFunctions/UpdateDevice.cs

```diff
@@ -40,42 +42,98 @@ public static class UpdateDevice
                 return new BadRequestObjectResult(Resources.InvalidIdMessage);
             }
 
+            // Parse out necessary data            
+            var mongo = MongoDefinition.MongoInstance;
+            IFormCollection reqForm;            
             Device device;
+            UpdateDeviceRequest data;
             try
             {
-                string requestBody = await new StreamReader(req.Body).ReadToEndAsync();
-                var data = JsonConvert.DeserializeObject<UpdateDeviceRequest>(requestBody);
-                device = new Device(
-                     ObjectId.Parse(id),
-                     ObjectId.Parse(data.DeviceTypeId),
-                     data.DeviceTag,
-                     data.Manufacturer,
-                     data.ModelNum,
-                     data.SerialNum,
-                     data.YearManufactured,
-                     data.Notes
-                 );
-
-                var reqBuildingId = data.Location.BuildingId;
-                ObjectId? buildingId = reqBuildingId != null ? ObjectId.Parse(reqBuildingId) : null;
-
-                device.SetLocation(
-                    buildingId,
-                    data.Location.Notes,
-                    data.Location.Latitude,
-                    data.Location.Longitude
-                );
-
-                device.SetFields(data.Fields);
-
-                device.SetLastModified(DateTime.UtcNow, "Anonymous");
-                // TODO: include device type attributes
+                reqForm = await req.ReadFormAsync();
+                data = JsonConvert.DeserializeObject<UpdateDeviceRequest>(reqForm["data"]);                              
+                
+                var deviceDocument = mongo.GetDevice(new ObjectId(id));
+                if (deviceDocument == null)
+                {
+                    return new NotFoundObjectResult(Resources.DeviceNotFoundMessage);
+                }
+
+                device = BsonSerializer.Deserialize<Device>(deviceDocument);                
             }
             catch (Exception)
             {
                 return new BadRequestObjectResult(Resources.CouldNotParseBody);
             }
 
+            if (!ObjectId.TryParse(data.DeviceTypeId, out _))
+            {
+                return new BadRequestObjectResult(Resources.InvalidIdMessage);
+            }            
+
+            // Set Data to new values
+            device.DeviceTypeId = ObjectId.Parse(data.DeviceTypeId);
+            device.DeviceTag = data.DeviceTag;
+            device.Manufacturer = data.Manufacturer;
+            device.ModelNum = data.ModelNum;
+            device.SerialNum = data.SerialNum;
+            device.YearManufactured = data.YearManufactured;
+            device.Notes = data.Notes;
+
+            var reqBuildingId = data.Location.BuildingId;
+            ObjectId? buildingId = reqBuildingId != null ? ObjectId.Parse(reqBuildingId) : null;
```

Comment:
Does the GetBsonDocument() method throw an exception if this is null?

Reply:
Yep it works, there is a test called ExistingDevice_UpdatesDeviceWithNoLocation_ReturnsOKResponse that verifies this.